packages/react-reconciler/src/ReactFiberNewContext.new.js

```diff
@@ -311,6 +329,112 @@ export function propagateContextChange<T>(
   }
 }
 
+export function lazilyPropagateParentContextChanges(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+) {
+  propagateParentContextChanges(current, workInProgress, renderLanes, true);
+}
+
+export function propagateParentContextChangesToDeferredTree(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+) {
+  propagateParentContextChanges(current, workInProgress, renderLanes, false);
+}
+
+function propagateParentContextChanges(
+  current: Fiber,
+  workInProgress: Fiber,
+  renderLanes: Lanes,
+  bailoutAfterFirstMatch: boolean,
+) {
+  if (!enableLazyContextPropagation) {
+    return false;
+  }
+
+  let parent = workInProgress;
+  while (parent !== null && (parent.flags & DidPropagateContext) === NoFlags) {
+    if (parent.tag === ContextProvider) {
+      const currentParent = parent.alternate;
+      if (currentParent !== null) {
+        const oldProps = currentParent.memoizedProps;
+        if (oldProps !== null) {
+          const providerType: ReactProviderType<any> = parent.type;
+          const context: ReactContext<any> = providerType._context;
+
+          const newProps = parent.pendingProps;
+          const newValue = newProps.value;
+
+          const oldValue = oldProps.value;
+
+          const changedBits = calculateChangedBits(context, newValue, oldValue);
+          if (changedBits !== 0) {
+            // The context value changed. Search for matching consumers and
+            // schedule them to update.
+            propagateContextChange(
+              workInProgress,
+              context,
+              changedBits,
+              renderLanes,
+            );
+            // If we found a matching consumer, we can stop any remaining
+            // propagation until the next level when we bail out.
+            if (
+              // propagateParentContextChangesToDeferredTree doesn't bail out
+              // after the first match; it propagates *all* the matching
+              // providers. This is used by Suspense and Offscreen because they
+              // spawn work in a subtree that isn't resumed until a later
+              // render, by which time we won't be able to tell that the
+              // provider has changed.
+              bailoutAfterFirstMatch &&
+              includesSomeLane(workInProgress.childLanes, renderLanes)
+            ) {
+              return;
+            }
+          }
+        }
+      }
+    }
+    parent = parent.return;
```

Comment:
guess I can comment on line items, not sure what I was seeing before. anyway this whole thing approach is very elegant. it fits in much cleaner than the way I did it

Reply:
Still very much based on your PR and RFC 😊